packages/next/build/webpack/loaders/eslint-loader/linter.ts

```diff
@@ -0,0 +1,232 @@
+import path from 'path'
+import { isAbsolute, join } from 'path'
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { writeFileSync, ensureFileSync } from 'fs-extra'
+import { interpolateName } from 'loader-utils'
+import ESLintError from './es-lint-error'
+import { CLIEngine, Linter as ESLinter, ESLint } from 'eslint'
+import { loader } from 'webpack'
+import createEngine from './create-engine'
+import { getBaseRules } from './base-rules'
+
+export class Linter {
+  public loaderContext: loader.LoaderContext
+  public options: any
+  private resourcePath: string
+  private engine: CLIEngine
+  private config: any
+  private isTypescript: Boolean = false
+
+  constructor(loaderContext: loader.LoaderContext, options: any) {
+    this.loaderContext = loaderContext
+    this.options = options
+    this.resourcePath = this.parseResourcePath()
+
+    // fixes for typescript
+    if (this.resourcePath.endsWith('ts') || this.resourcePath.endsWith('tsx')) {
+      this.isTypescript = true
+      if (!options.parserOptions.plugins.includes('typescript')) {
+        options.parserOptions.plugins.push('typescript')
+      }
+    } else {
+      options.parserOptions.babelOptions = options.parserOptions
+        .babelOptions || {
+        parserOpts: {
+          plugins: ['estree', 'jsx'],
+        },
+      }
+      options.parserOptions.requireConfigFile =
+        options.parserOptions.requireConfigFile || false
+    }
+    options.baseConfig = {
+      extends: ['eslint:recommended', 'plugin:@next/next/recommended'],
+      rules: getBaseRules(this.isTypescript),
+      plugins: ['import', 'react-hooks'],
+    }
+    options.parser = this.isTypescript
+      ? require.resolve('@typescript-eslint/parser')
+      : require.resolve('@babel/eslint-parser')
+    const { engine } = createEngine(options)
+    this.engine = engine
```

Comment:
Do we need to watch external `.eslintrc` files and show a message for the dev server to be restarted for the changes to take affect or can we hot-reload external rules?

Reply:
is this a thing we do for TS config? if yes, can you point to the example where its done

Reply:
We don't do this for `tsconfig` yet ([related issue](https://github.com/vercel/next.js/issues/13964)) so definitely doesn't need to be done in this PR, was just wondering if there was already an API we could leverage here to achieve this for `.eslintrc` changes, if not we can address this when the other issue is addressed.